File: packages/node-sdk/src/log.ts
/**
 * Structured logging for the Neon3 SDK.
 *
 * Usage:
 * ```ts
 * import { configure, getLogger } from "@neon3/sdk/log";
 * configure("debug");
 * const log = getLogger("rpc");
 * ```
 *
 * Honors the `NEON3_LOG_LEVEL` env var (debug/info/warn/error) on first call.
 */

let configured = false;
type Level = "debug" | "info" | "warn" | "error";
let level: Level = "warn";

const ORDER: Record<Level, number> = { debug: 10, info: 20, warn: 30, error: 40 };

export function configure(l: Level | string = "warn"): void {
  if (configured) return;
  level = (l as Level) in ORDER ? (l as Level) : "warn";
  const env = (process.env.NEON3_LOG_LEVEL as string | undefined)?.toLowerCase();
  if (env && env in ORDER) level = env as Level;
  configured = true;
}

export function getLogger(name = ""): {
  debug: (msg: string, ...args: unknown[]) => void;
  info: (msg: string, ...args: unknown[]) => void;
  warn: (msg: string, ...args: unknown[]) => void;
  error: (msg: string, ...args: unknown[]) => void;
} {
  if (!configured) configure();
  const prefix = name ? `neon3/${name}` : "neon3";
  const emit = (lvl: Level, msg: string, args: unknown[]) => {
    if (ORDER[lvl] < ORDER[level]) return;
    const ts = new Date().toISOString();
    // eslint-disable-next-line no-console
    console.error(`${ts} ${lvl.toUpperCase().padEnd(5)} ${prefix}: ${msg}`, ...args);
  };
  return {
    debug: (m, ...a) => emit("debug", m, a),
    info: (m, ...a) => emit("info", m, a),
    warn: (m, ...a) => emit("warn", m, a),
    error: (m, ...a) => emit("error", m, a),
  };
}

File: packages/python-sdk/src/neon3_sdk/log.py
"""Structured logging for the Neon3 SDK.

Usage::

    import logging
    from neon3_sdk.log import get_logger, configure

    configure(level=logging.DEBUG)  # or NEON3_LOG_LEVEL=debug env var
    log = get_logger("render")

Every RPC call logs one line at INFO: target, method, request_id, elapsed_ms,
ok/err. Set level to DEBUG for frame-level detail.
"""

from __future__ import annotations

import logging
import os
import sys

_ROOT_NAME = "neon3_sdk"

_configured = False


def configure(level: int | str = logging.INFO) -> None:
    """Configure the neon3_sdk logger. Idempotent."""
    global _configured
    if _configured:
        return
    if isinstance(level, str):
        level = getattr(logging, level.upper(), logging.INFO)
    logger = logging.getLogger(_ROOT_NAME)
    logger.setLevel(level)
    if not logger.handlers:
        handler = logging.StreamHandler(stream=sys.stderr)
        handler.setFormatter(logging.Formatter(
            "%(asctime)s %(levelname)-7s %(name)s: %(message)s"
        ))
        logger.addHandler(handler)
    logger.propagate = False
    _configured = True


def get_logger(name: str = "") -> logging.Logger:
    """Return a child logger under ``neon3_sdk``.

    Honors the ``NEON3_LOG_LEVEL`` env var (``debug``/``info``/``warning``/
    ``error``) on first call.
    """
    if not _configured:
        env_level = os.environ.get("NEON3_LOG_LEVEL", "warning")
        configure(env_level)
    if name:
        return logging.getLogger(f"{_ROOT_NAME}.{name}")
    return logging.getLogger(_ROOT_NAME)
